File: algorithms/code/leetcode/lc171_excel_sheet_column_number/lc171_excel_sheet_column_number.py
# https://leetcode.com/problems/excel-sheet-column-number/

import unittest


class Solution:
    def titleToNumber(self, columnTitle: str) -> int:
        s = columnTitle
        if not s:
            return -1
        result = 0
        i = 0
        while i < len(s):
            result *= 26
            result += ord(s[i]) - ord('A') + 1
            i += 1
        return result


class TestCode(unittest.TestCase):
    def test_A(self) -> None:
        assert Solution().titleToNumber('A') == 1

    def test_B(self) -> None:
        assert Solution().titleToNumber('B') == 2

    def test_C(self) -> None:
        assert Solution().titleToNumber('C') == 3

    def test_Z(self) -> None:
        assert Solution().titleToNumber('Z') == 26

    def test_AA(self) -> None:
        assert Solution().titleToNumber('AA') == 27

    def test_AB(self) -> None:
        assert Solution().titleToNumber('AB') == 28

    def test_nothing(self) -> None:
        assert Solution().titleToNumber('') == -1

File: algorithms/code/leetcode/lc171_excel_sheet_column_number/lc171_excel_sheet_column_number.rs
// https://leetcode.com/problems/excel-sheet-column-number/
// #easy

pub struct Solution;

impl Solution {
    pub fn title_to_number(s: String) -> i32 {
        if s.is_empty() {
            return -1;
        }
        let mut result = 0;
        for c in s.chars() {
            result *= 26;
            result += c as i32 - 'A' as i32 + 1;
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_a() {
        assert_eq!(Solution::title_to_number("A".into()), 1);
    }

    #[test]
    fn test_b() {
        assert_eq!(Solution::title_to_number("B".into()), 2);
    }

    #[test]
    fn test_c() {
        assert_eq!(Solution::title_to_number("C".into()), 3);
    }

    #[test]
    fn test_z() {
        assert_eq!(Solution::title_to_number("Z".into()), 26);
    }

    #[test]
    fn test_aa() {
        assert_eq!(Solution::title_to_number("AA".into()), 27);
    }

    #[test]
    fn test_ab() {
        assert_eq!(Solution::title_to_number("AB".into()), 28);
    }

    #[test]
    fn test_nothing() {
        assert_eq!(Solution::title_to_number("".into()), -1);
    }
}
